### PR TITLE
DebugPrintf: various cleanups and add documentation

### PR DESCRIPTION
**Pull request by nhaehnle**
Series of separate commits around cleaning up DebugPrintf support.

One unrelated addition is the patch "lgc: add initial pipeline state printer and recorder facilities" whose purpose is to improve the testability of the LowerDebugPrintf pass and potentially other passes.

Note that this depends on https://github.com/GPUOpen-Drivers/llvm-dialects/pull/54 (merged)

### CODE REVIEW

**Review by jasilvanus — COMMENTED** [review]
Just took a look at the printf part.

**Review by Flakebi — COMMENTED** [review]
LGTM, two nits inline

File: lgc/interface/lgc/LgcDialect.td [hunk 1]
```diff
@@ -81,6 +82,19 @@ def BufferPtrDiffOp : LgcOp<"buffer.ptr.diff", [Memory<[]>, WillReturn]> {
   }];
 }
 
+def DebugPrintfOp : LgcOp<"debug.printf", [Memory<[(readwrite InaccessibleMem)]>, WillReturn]> {
+  let arguments = (ins BufferPointer:$buffer, ConstantPointer:$format, varargs:$args);
+  let results = (outs);
+
+  let summary = "print a formatted message";
+  let description = [{
+    Writes an entry to the debug printf buffer pointed to by `buffer`. No-op if `buffer` is literal constant poison.
```
[the remark this answers is not in the document]
Comment (nhaehnle) [reply lgc/interface/lgc/LgcDialect.td:91]: The reasoning isn't great, but it has to do with resource nodes being unavailable in SPIRVReader, so the SPIRVReader can't tell whether the debug buffer is present or not.

File: lgc/interface/lgc/LgcDialect.td [hunk 2]
```diff
@@ -81,6 +82,19 @@ def BufferPtrDiffOp : LgcOp<"buffer.ptr.diff", [Memory<[]>, WillReturn]> {
...
+
+    `format` must be a non-external global variable in the constant address space of type `[n x i8]` with a
+    nul-terminated initializer.
```
[the remark this answers is not in the document]
Comment (nhaehnle) [reply lgc/interface/lgc/LgcDialect.td:94]: Oh, actually that's outdated -- we don't expect null-terminators here. Will change.

**Review by Flakebi — APPROVED** [review]

**Comment by amdvlk-admin** [comment]
### Test summary for commit 026d20bbf397b15dc49a5b72f6846b36616bed8c
<details><summary>CTS tests (Failed: 0/140867)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39059/70432 (55.5%)</li>
<li>Failed: 0/70432 (0.0%)</li>
<li>Not Supported: 31373/70432 (44.5%)</li><li>Warnings: 0/70432 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39090/70435 (55.5%)</li>
<li>Failed: 0/70435 (0.0%)</li>
<li>Not Supported: 31345/70435 (44.5%)</li><li>Warnings: 0/70435 (0.0%)</li>
</ul></details>


</ul></details>

**Review by jiaolu — APPROVED** [review]

**Review by amdrexu — APPROVED** [review]

**Pull request merged**